File: Stock.java
import java.io.*;
import java.net.*;
import java.util.*;

/*
    APIs Used:
        Alpha Vantage: https://www.alphavantage.co/documentation/
 */

public class Stock {

    protected ArrayList<String[]> RawData = new ArrayList<>();
    protected ArrayList<String[]> DayData = new ArrayList<>();
    protected ArrayList<ArrayList<String[]>> WeekData = new ArrayList<>(); //Day,Data Select
    protected String Interval;
    protected String Ticker;
    protected String Slice;
    protected boolean forSMVI;
    protected String Time_Series = "INTRADAY_EXTENDED";
    final String APIKey = "4HRPSUDJ4S8WR99F";
    protected double LatestOpeningPrice;
    protected double SevenDayOpeningPrice;

    public Stock(String Ticker, String Interval, String Slice, boolean forSMVI) throws IOException {
        this.Interval = Interval;
        this.Ticker = Ticker;
        this.Slice = Slice;
        this.forSMVI = forSMVI;
        if (Interval.equals("")) this.Interval = "5";
        if (Ticker.equals("")) this.Ticker = "IBM";
        if (Slice.equals("")) this.Slice = "year1month1";
        run();
    }

    /**
     * Used to test the Stock class
     */
    public Stock() {
        Interval = "5";
        Ticker = "IBM";
        Slice = "year1month1";
        try {
            run();
        } catch (Exception e) {
            System.out.println(e);
        }
    }

    public void run() throws IOException {
        getHistory();
    }

    private void getHistory(){
        Interval += "min";
        Time_Series = "TIME_SERIES_" + Time_Series;
        try{
            URL url = new URL("https://www.alphavantage.co/query?function="+Time_Series+"&symbol="+Ticker+"&interval="+Interval+"&slice="+Slice+"&apikey="+APIKey);
            HttpURLConnection urlConnection = (HttpURLConnection) url.openConnection();
            InputStream in = url.openStream();
            BufferedReader reader = new BufferedReader(new InputStreamReader(in));
            File file;
            if(forSMVI) file = new File("Data//DIA_Data.csv");
            else file = new File("Data//Data.csv");
            FileWriter writer = new FileWriter(file);
            String line;
            while((line = reader.readLine()) != null) {
                writer.write(line+"\n");
            }
            writer.close();
            reader.close();
            urlConnection.disconnect();
            RawData = parseData();
            DayData = getDayData();
            WeekData = GetWeekData();
            LatestOpeningPrice = getLatestOpenPrice();
            SevenDayOpeningPrice = getSevenDayOpeningPrice();
        } 
        catch(IOException E){
            System.out.println("Error in \"getHistory\"\n"+E);}
    }

    private ArrayList<String[]> parseData() throws IOException {
        ArrayList<String[]> ParsedData = new ArrayList<String[]>();
        BufferedReader reader = new BufferedReader(new FileReader("Data\\Data.csv"));
        String line;
        while((line = reader.readLine()) != null){
            ParsedData.add(line.split(","));
        }
        reader.close();
        return ParsedData;
    }

    private ArrayList<String[]> getDayData() {
        ArrayList<String[]> DailyData = new ArrayList<>();
        DailyData.add(RawData.get(0));
        String CurrentDay = RawData.get(1)[0].substring(0, RawData.get(1)[0].indexOf(":") - 3);
        for (int i = 1; i < RawData.size(); i++) {
            if (RawData.get(i)[0].contains(CurrentDay)) DailyData.add(RawData.get(i));
            else break;
        }
        return DailyData;
    }

    private ArrayList<String[]> getDayData(String Date){
        ArrayList<String[]> DayData = new ArrayList<>();
        int start = -1;
        int end = -1;
        for(int i = 1; i < RawData.size(); i++){
            if(RawData.get(i)[0].substring(0,RawData.get(i)[0].indexOf(":")-3).equals(Date.substring(0,Date.indexOf(":")-3))){
                if(start == -1) start = i;
                end = i;
            }
        }

        for(int f = start; f <= end; f++){
            DayData.add(RawData.get(f));
        }
        return DayData;
    }

    private ArrayList<ArrayList<String[]>> GetWeekData(){
        ArrayList<ArrayList<String[]>> WeekData = new ArrayList<>();
        String[] Week = new String[7];
        Week[0] = RawData.get(1)[0];
        int Index = 1;
        for(int i = 2; i < RawData.size(); i++){
            if(Index > 6) break;
            if(!RawData.get(i)[0].substring(0,RawData.get(i)[0].indexOf(":")-3).equals(Week[Index-1].substring(0,Week[Index-1].indexOf(":")-3))){
                Week[Index] = RawData.get(i)[0];
                Index++;
            }
        }
        for(String date : Week){
            WeekData.add(getDayData(date));
        }
        return WeekData;
    }

    private double getLatestOpenPrice(){
        return Double.parseDouble(DayData.get(DayData.size()-1)[1]);
    }

    private double getSevenDayOpeningPrice(){ return Double.parseDouble(WeekData.get(6).get(WeekData.get(6).size()-1)[1]);}

    /**
     * A method to find the average price of a stock on a set interval on the most recent day
     *
     * @return the average price of a stock as a double
     */
    public double avgDayPrice() {
        double tp = 0.0;
        for (int i = 1; i < DayData.size(); i++) {
            tp += (Double.parseDouble(RawData.get(i)[1]) + Double.parseDouble(RawData.get(i)[2]) + Double.parseDouble(RawData.get(i)[3]) + Double.parseDouble(RawData.get(i)[4])) / 4;
        }
        double ap = tp / (DayData.size() - 1);
        return ap;
    }

    /**
     * A method to find the average price of a stock on a given day
     *
     * @param data refers to the 2d array containing this format [Time, Open, High, Low, Close, Volume]
     * @return the average price of a given stock as a double
     */
    public double avgDayPrice(ArrayList<String[]> data) {
        double tp = 0.0;
        for (int i = 0; i < data.size(); i++) {
            tp += (Double.parseDouble(data.get(i)[1]) + Double.parseDouble(data.get(i)[2]) + Double.parseDouble(data.get(i)[3]) + Double.parseDouble(data.get(i)[4])) / 4;
        }
        double ap = tp / (data.size() - 1);
        return ap;
    }

    public double avgSevenDayPrice() {
        double total = 0;
        for (ArrayList<String[]> as : WeekData) {
            total += avgDayPrice(as);
        }
        return total / 7;
    }

    public double avgSevenDayPrice(ArrayList<ArrayList<String[]>> weekdata) {
        double total = 0;
        for (ArrayList<String[]> as : weekdata) {
            total += avgDayPrice(as);
        }
        return total / 7;
    }

    public double avgPrice(ArrayList<ArrayList<String[]>> c){
        double total = 0;
        for (ArrayList<String[]> as : c) {
            total += avgDayPrice(as);
        }
        return total / c.size();
    }

    /**
     * A method to find the standard deviation of a given stock for the most recent day
     *
     * @return the standard deviation of the stock as a double
     */
    public double DayDeviation() {
        double top = 0.0;
        for (int i = 1; i < DayData.size(); i++) {
            top += Math.pow(((Float.parseFloat(DayData.get(i)[1]) + Float.parseFloat(DayData.get(i)[2]) + Float.parseFloat(DayData.get(i)[3]) + Float.parseFloat(DayData.get(i)[4])) / 4.0) - avgDayPrice(), 2.0);
        }
        double bottom = DayData.size() - 2;
        double total = top / bottom;
        return Math.sqrt(total);
    }

    public double sevenDayDeviation(){
        double top = 0.0;
        for(int i = 0; i < 7; i++){
            for(ArrayList<String[]> as : WeekData){
                top += Math.pow( avgDayPrice(as)-avgSevenDayPrice(),2);
            }
        }
        double Bottom = 0.0;
        for(int i = 0; i < 7; i++){
            Bottom += WeekData.get(i).size()-1;
        }
        Bottom--;
        return Math.sqrt(top/Bottom);
    }

    public double deviation(ArrayList<String[]> a){
        //Check if the arraylist has not been formatted correctly, meaning it still has headers
        try{
            double d = Double.parseDouble(a.get(0)[0]);
        }catch(Exception e){ a.remove(0);}
        ////////////////////////////////////////////////////////////////////////////////////////
        double top = 0.0;
        double avg = avgDayPrice(a);
        double tmavg = 0.0;
        for(String[] sa : a){
            tmavg += Double.parseDouble(sa[2])+Double.parseDouble(sa[3]);
        }
        tmavg = tmavg / (2*a.size());
        for(String[] sa: a){
            top += (Double.parseDouble(sa[2])+Double.parseDouble(sa[3]))-tmavg;
        }
        double bottom = a.size()-1;
        return Math.sqrt(top/bottom);
    }

    public double getVariation(ArrayList<String[]> data){
        return Math.pow(deviation(data),2);
    }

    public double currentDayDeviationPercentage() {
        return DayDeviation()/avgDayPrice();
    }

    public double sevenDayDeviationPercentage(){
        return sevenDayDeviation()/avgSevenDayPrice();
    }

}

File: Stock.py
from FinnHub import FinnHub
import PySimpleGUI
import datetime
import math
import csv
import os

class Stock:
    def __init__(self,Ticker,Interval_Index,StartSlice,EndSlice,AllSlices,forSMVI):
        self.Ticker = Ticker.upper()
        self.AllIntervals = [1, 5, 15, 30, 60, 'D', 'W', 'M']
        self.Interval = self.AllIntervals[Interval_Index]
        self.StartSlice = StartSlice
        self.EndSlice = EndSlice
        self.AllSlices = AllSlices
        self.forSMVI = forSMVI
        self.getTickers()
        self.errorHandling()
        if not self.hasErrors: self.run()

    def getTickers(self):
        self.allTickers = []
        with open(os.getcwd() + '/Data/tickers.csv') as file:
            reader = csv.reader(file)
            for line in reader:
                if line[0] != 'Symbol': self.allTickers.append(line[0])

    def popup(self,text):
        PySimpleGUI.Popup(text)
        self.hasErrors = True

    def errorHandling(self):
        self.hasErrors = False
        if self.Ticker == 'Ticker':
            self.popup('Error, the asset ticker that has been given has an error')
        elif self.Interval not in self.AllIntervals:
            self.popup('Error, the interval that has been given is incorrect')
        elif self.StartSlice not in self.AllSlices:
            self.popup('Error, the starting time that has been given is incorrect')
        elif self.EndSlice not in self.AllSlices:
            self.popup('Error, the ending time that has been given is incorrect')
        elif self.AllSlices.index(self.EndSlice) < self.AllSlices.index(self.StartSlice):
            self.popup('Error, the ending time is before the starting time')
        elif not self.forSMVI and self.Ticker not in self.allTickers:
            self.getTickers()
            if self.Ticker not in self.allTickers:
                self.popup('Error, the given ticker is invalid')

    def toDate(self,foo,noTime):
        if noTime: return datetime.datetime.fromtimestamp(foo).strftime('%Y-%m-%d')
        return datetime.datetime.fromtimestamp(foo).strftime('%Y-%m-%d %H:%M:%S')

    #Fee fi fo fum, parseToArray to get things done
    def parseToArray(self,fee):
        return [fo.split(',') for fo in fee]

    def getAverage(self,foo):
        Open = float(foo[1])
        Close = float(foo[4])
        High = float(foo[2])
        Low = float(foo[3])
        return (Open + Close + High + Low)/4

    def addArrays(self):
        self.headers = self.finnHub.headers
        self.Times = self.Open = self.High = self.Low = self.Close = self.Volume = []
        for line in self.parseToArray(self.finnHub.allData):
            self._extracted_from_addArrays_5(line)
        self.allData = {
                        'Times'  :   self.Times,
                        'Open'   :   self.Open,
                        'High'   :   self.High,
                        'Low'    :   self.Low,
                        'Close'  :   self.Close,
                        'Volume' :   self.Volume
                        }

    def _extracted_from_addArrays_5(self, line):
        self.Times.append(float(line[0]))
        self.Open.append(float(line[1]))
        self.High.append(float(line[2]))
        self.Low.append(float(line[3]))
        self.Close.append(float(line[4]))
        self.Volume.append(float(line[4]))

    # data : a 2D array where each entry is an array, which follows the following format
    # [Times,Open,High,Low,Close,Volume]
    def standardDeviation(self,data):
        N = len(data)
        if N == 0:
            self.popup('Unknown Error, please contact daniel_philips@asl.org')
        tempMu = 0
        Points = []
        for time in data: 
            tempMu += self.getAverage(time)
            Points.append(self.getAverage(time))
        Mu = tempMu/N
        top = sum((xi - Mu)**2 for xi in Points)
        return math.sqrt( top / N )
        
    def run(self):
        self.finnHub = FinnHub(self.StartSlice,self.EndSlice,self.Ticker,self.Interval,self.forSMVI)
        self.addArrays()
        self.SDToDisplay = self.standardDeviation(self.finnHub.allData)